### PR TITLE
Add expiration to FunctionOfTime::PiecewisePolynomial

### PR DESCRIPTION
**Pull request by markscheel**
## Proposed changes

The expiration time will be used to tell whether the
FunctionsOfTime in the GlobalCache are up to date.

Interface changes:
 - The one nontrival constructor of PiecewisePolynomial now takes
   an additional value expiration_time.
 - The update function of PiecewisePolynomial now takes an
   expiration_time.
 - PiecewisePolynomial::time_bounds() now returns expiration_time as
   its upper bounds.
 - FunctionOfTime::time_bounds() documentation now states that the
   bounds includes any allowed time-extrapolation.
 - Many tests have been changed to accomodate the interface changes
   in PiecewisePolynomial; this accounts for the vast majority of
   file changes in this commit.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

My current model is to use the upper bound of FunctionOfTime::time_bounds() to determine whether a FunctionOfTime is out of date. Alternatively I could add a new member function FunctionOfTime::expiration_time() expressly for that purpose, but I chose to keep fewer member functions in the interface.

### CODE REVIEW

File: src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp
Comment (wthrowe) [inline src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp:85]: Should there also be a requirement that `time_of_update >= expiration_time_`?  It seems like you would get race conditions if you updated before a time that could have alredy been used.

File: src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp
Comment (wthrowe) [inline src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp]: I'm confused by this one.  It looks like `func` is a `std::array<DataVector, ...>`, which should support move assigning to an element.

**Review by wthrowe — CHANGES_REQUESTED** [review]

**Comment by kidder** [comment]
I am confused by what is going on in this PR.  Why is the `expiration_time` being initialized to the `initial_time` in various constructors?  Where is the `ERROR` if I evaluate `func` or `func_and_deriv`, etc. past the `expiration_time`?

File: src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp
Comment (markscheel) [reply src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp]: I had changed this line, and I had simultaneously changed the type of the input argument updated_max_deriv (used to be a DataVector passed by value and I changed it to a `const DataVector &`) because clang-tidy complained about passing by value.  If you want, I can restore the std::move, restore the pass-DataVector-by-value, and add a NOLINT for clang-tidy.

Comment (wthrowe) [reply src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp]: Yeah, I have no idea what clang-tidy is thinking here.  I can reproduce the warnings locally, and it is literally warning that the variable should be made const, and when you do that warning that the variable should not be const.  Just find somewhere to put a NOLINT to shut it up.

**Comment by markscheel** [comment]
I did the following:

 - rebased on develop (and did various resulting new clang-tidy fixes in extra fixup commits)
 - pushed fixups to address review of @wthrowe 
 - Fixups: Added ERROR if `func_and_deriv(t)` is called with t past the expiration time (and a test)
 - Fixups: Added a few more ERROR conditions (and tests for them)
 - Fixups: Re-did all the tests so that they don't set expiration_time to initial_time in the constructors, and instead choose some appropriate expiration_time.
 - Fixups: Allow FunctionOfTimeUpdater to update the expiration time if it decides that the FunctionOfTime is ok to not be updated until later.
 - Fixups: Add new option to the 3 DomainCreators that create a PiecewisePolynomial; this option sets the initial expiration time.  And then set this option in the various tests that use these DomainCreators.

Some of my new fixups (particularly the ones that set expiration_time in the tests) redo some of my earlier commits and touch the same code in various files.  Therefore I can reduce the size of this PR a lot if I squash.  Should I squash?

File: src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp
Comment (wthrowe) [inline src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp]: Use `NOLINT(performance-unnecessary-value-param)` (or the `NOLINTNEXTLINE` version).

File: src/Domain/FunctionsOfTime/PiecewisePolynomial.hpp
Comment (wthrowe) [inline src/Domain/FunctionsOfTime/PiecewisePolynomial.hpp]: This should have a check that the reset doesn't move the time backwards.

**Review by wthrowe — CHANGES_REQUESTED** [review]
These and the previous fixups can be squashed as far as I am concerned.

**Comment by kidder** [comment]
yes, please squash and then I will take a more careful look

**Comment by markscheel** [comment]
Squashed.

**Comment by markscheel** [comment]
Oh hold on, I want to move a few changes from one commit to another...

**Comment by markscheel** [comment]
Ok, attempted to rearrange files in each commit to make things easier to understand.

**Review by wthrowe — DISMISSED** [review]

**Comment by markscheel** [comment]
Fixed failing test (problem was a `<=` in an error check that should have been a `<`).

File: src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp
Comment (wthrowe) [inline src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp:85]: Either "<=" -> "<" or "nondecreasing" -> "increasing".

The current check does not match the one in `reset_expiration_time`.  I don't know if that is intentional.

**Review by wthrowe — CHANGES_REQUESTED** [review]

**Comment by markscheel** [comment]
Made change suggested by @wthrowe and squashed (since the change was one character)

**Review by wthrowe — DISMISSED** [review]

**Comment by markscheel** [comment]
The failing test was a timeout in tests that I didn't touch, like Swsh. Restarted the tests.

File: src/Domain/Creators/TimeDependence/CubicScale.hpp
Comment (nilsdeppe) [inline src/Domain/Creators/TimeDependence/CubicScale.hpp]: I think you should add that `If the function or expiration time is not updated by the end of the interval the simulation will fail.` because you haven't really supplied any way of saying "It never gets updated." Would this be a good place for using `Auto`?

File: src/Domain/FunctionsOfTime/PiecewisePolynomial.hpp
Comment (kidder) [inline src/Domain/FunctionsOfTime/PiecewisePolynomial.hpp]: expiration_time - next_expiration_time

File: src/Domain/FunctionsOfTime/PiecewisePolynomial.hpp
Comment (kidder) [inline src/Domain/FunctionsOfTime/PiecewisePolynomial.hpp:61]: expiration_time -> next_expiration_time

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash

File: src/Domain/Creators/TimeDependence/CubicScale.hpp
Comment (markscheel) [reply src/Domain/Creators/TimeDependence/CubicScale.hpp]: Yes, I agree that UpdateDeltaT is not optimal, partially because there's no way of saying "It never gets updated". Furthermore, I don't like the need to specify UpdateDeltaT here at all, because expiration times should be set by the control system rather than by the DomainCreator constructors.   However, the `FunctionsOfTime::PiecewisePolynomial` needs an expiration time in its constructor, and that constructor  is called by `CubicScale::functions_of_time()`.  Note that `CubicScale::functions_of_time()` is the only place `UpdateDeltaT` is used.

I can see a few possibilities to improve this situation. Either:

 - Rename `UpdateDeltaT` to `InitialExpirationTime` since the option is used only for initializing the `PiecewisePolynomial`. Then use `Options::Auto` to set the automatic value to  `std::numeric_limits<double>::max()` (a.k.a. "it never gets updated"). Rely on the control system to reset the expiration time as needed.
 - Eliminate the option entirely, and set the initial expiration time to `std::numeric_limits<double>::max()` always, expecting that the control system will reset the expiration time as needed.
 
Both of the above  possibilities (as stated) would require me to remove the `next_expiration_time >= expiration_time_` check in `PiecewisePolynomial::reset_expiration_time` (because if the initial expiration time is `std::numeric_limits<double>::max`, the check would always fail).  If we still want the check in `PiecewisePolynomial::reset_expiration_time`, we could do either:
 - Set the initial expiration time to `std::numeric_limits<double>::lowest()` so that the `PiecewisePolynomial` is initially invalid, and then rely on the control system to reset it, including before the first time it gets used.
 - Set the initial expiration time to `initial_time`, so the `PiecewisePolynomial` is initially valid only for its initial time.  Rely on the control system to reset it before it is used for any subsequent times.
 - Change `PiecewisePolynomial::expiration_time_` from a double to a `std::optional<double>`, initially default-constructed, and interpret an empty `expiration_time_` as "never expires". Allow `PiecewisePolynomial::reset_expiration_time` to reset an empty `expiration_time_` to anything.

Thoughts?

Comment (markscheel) [reply src/Domain/Creators/TimeDependence/CubicScale.hpp]: I currently favor eliminating the option entirely and setting the initial expiration_time to initial_time, because I don't know how the DomainCreator is going to know how to set the initial expiration time.

Comment (markscheel) [reply src/Domain/Creators/TimeDependence/CubicScale.hpp]: Consensus from SpECTRE meeting Nov 23, 2020: Rename `UpdateDeltaT` -> `InitialExpirationDeltaT`, add `Auto` to this option so that `Auto` makes `expiration_time` become `std::numeric_limits<double>::max()` (i.e. never expires), which is never resettable or updatable.

**Comment by markscheel** [comment]
Made changes, squashed @kidder 's suggested changes, did not squash last commit that does renaming and Auto.

**Review by nilsdeppe — COMMENTED** [review]
I like the `Auto` change and the rename :+1: Thanks for doing that!

I'm happy for you to squash. No need to wait for my approval. I only glanced over the PR to begin with anyway :)

**Comment by markscheel** [comment]
Squashed

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
@wthrowe please look at

**Review by wthrowe — APPROVED** [review]

**Pull request merged**
